File: src/com/designpatterns/creational/factory/ShapeAnalyzer.java
package com.designpatterns.creational.factory;

public class ShapeAnalyzer {

    private final IShapeFactory shapeFactory;

    public ShapeAnalyzer(IShapeFactory shapeFactory) {
        this.shapeFactory = shapeFactory;
    }

    public void analyzeShapes() {
        IShape circularShape = shapeFactory.getShape(Circle.TAG);
        IShape rectangularShape = shapeFactory.getShape(Rectangle.TAG);
        IShape squareShape = shapeFactory.getShape(Square.TAG);

        circularShape.draw();
        rectangularShape.draw();
        squareShape.draw();
    }
}

File: src/com/designpatterns/creational/factory/kotlin/ShapeAnalyzer.kt
package com.designpatterns.creational.factory.kotlin

class ShapeAnalyzer(private val shapeFactory: IShapeFactory) {

    fun analyzeShapes() {
        // Get shapes
        val circularShape = shapeFactory.getShape(Circle.TAG)
        val rectangularShape = shapeFactory.getShape(Rectangle.TAG)
        val squareShape = shapeFactory.getShape(Square.TAG)

        // Draw shapes
        circularShape.draw()
        rectangularShape.draw()
        squareShape.draw()
    }
}